feat(ivy): Add property instruction

benlesh opened the pull request:
Adds a minimal change, introducing  a `select` instruction, and a `property` instruction, per [this design document](https://hackmd.io/c/BkDUxaW84/%2Fdi2r3ZeuSQ284ttBUPoLBg)

mhevery commented on packages/core/src/render3/instructions/instructions.ts:
````diff
@@ -1141,6 +1141,60 @@ export function elementAttribute(
+ * ```
+ * @param index the index of the item to act on with the following instructions
+ */
+export function select(index: number) {
````
Could we start putting these into separate files. `instructions/select`. Also could we get the name aligned with https://hackmd.io/s/SJyvHRwuN

benlesh replied:
I would like to do the moving and renaming in a separate PR, if that's okay.

mhevery replied:
SGTM

benlesh replied:
Discussed offline: Agreed to do it later.

kara commented on packages/core/test/render3/instructions_spec.ts:
```diff
@@ -162,6 +163,93 @@ describe('instructions', () => {
     });
   });
 
+  describe('select', () => {
```
Hmm I really don't love that we're adding new tests with generated code, since we agreed we would no longer do this. It's adding tech debt. Can we add a big TODO comment here to remove it when the compiler part is added?  And promise you will migrate these as soon as we have that?

mhevery replied:
Yes, let's not do this, instead let's just update the compiler to use the new instructions.

benlesh replied:
Discussed offline: Added `TODO`s to the tests we need to remove once this is wired into the compiler. But we'll want to leave the tests that are checking to make sure assertions throw in dev mode if things get in a bad state.

mhevery reviewed (changes requested):
Yes you are on the right track. See comments.

mhevery commented on packages/core/src/render3/instructions/instructions.ts:
```diff
@@ -1141,6 +1141,60 @@ export function elementAttribute(
+ * (e.g. `property('name', ctx.name)('title', ctx.title)`)
+ */
+export function property<T>(
+    propName: string, value: T, sanitizer?: SanitizerFn | null,
```
Not sure what you mean. The optional args would need to stay.

kara replied:
Discussed offline. For the curious: we have discussed trying to avoid optional args because they cause intermediate functions to be created, but we should probably do this cleanup at a later time.

benlesh replied:
Leaving them for this PR, then.

benlesh committed:
feat(ivy): add property instruction

- Adds `property` instruction
- Does _NOT_ add compiler changes to accommodate `property` instruction, that will be a follow up PR.
- Updates `select` instruction to set the selected index in state.
- Adds dev mode assertions around the selected index state.

Related #29527

benlesh commented:
This is technically blocked by #29546 at this point, as it will build on top of that.

benlesh commented on packages/core/src/render3/instructions/instructions.ts:
````diff
@@ -1141,6 +1141,60 @@ export function elementAttribute(
+ * ```
+ * @param index the index of the item to act on with the following instructions
+ */
+export function select(index: number) {
````
Right on. That work is now in with #29527 and more after that with #29546

kara commented on packages/core/test/render3/instructions_spec.ts:
```diff
@@ -162,6 +163,95 @@ describe('instructions', () => {
+
+    // TODO(benlesh): Replace with TestBed tests once the instruction is being generated.
+    it('should chain', () => {
+      const t = new TemplateFixture(createDiv, () => {}, 1, 2);
```
Can you add a comment here to show what the user would write?

```ts
// <div [title]="title" [accesskey]="key"></div>
```

benlesh replied:
Done for each test.

kara reviewed (approved):
LGTM

benlesh commented:
[presubmit](https://test.corp.google.com/ui#id=OCL:240598077:BASE:240629026:1553718502839:e4954475)

kara commented on packages/core/src/render3/instructions/instructions.ts:
```diff
@@ -1142,6 +1169,34 @@ export function elementAttribute(
+ */
+export function property<T>(
+    propName: string, value: T, sanitizer?: SanitizerFn | null, nativeOnly?: boolean):
+    any {  // NOTE: `any` returned here because tsickle freaked out over recursive return
```
Let's make this a TODO with the relevant issue linked

```suggestion
    any {  // TODO: Update from `any` when tsickle bug with recursive returns is fixed (https://github.com/angular/tsickle/issues/1009)
```

benlesh replied:
You're too quick! :) Done... but moved up to the type that @mhevery suggested.

kara reviewed (approved):
Looks good, pending comment change

kara commented on packages/core/src/render3/instructions/instructions.ts:
```diff
@@ -1141,7 +1168,42 @@ export function elementAttribute(
 
+// TODO: Remove this when the issue is resolved.
+/**
+ * Tsickle freaked out over a function returning itself
```
Can we be a bit more precise than "freaked out"? 

```suggestion
 * Tsickle has a bug where it creates an infinite loop for a function returning itself.
 * This is a temporary type that will be removed when the issue is resolved.
```

kara committed:
refactor: update tsickle comment to be more precise

Co-Authored-By: benlesh <ben@benlesh.com>

benlesh commented:
The new [presubmit](https://test.corp.google.com/ui#id=OCL:240598077:BASE:240670689:1553730359234:970ba2db) appears to have unrelated errors.
